app/adapter/db/availablekey.go

```diff
@@ -18,93 +14,37 @@ type AvailableKeySQL struct {
 }
 
 // Create inserts a new key into available_key table.
-func (a AvailableKeySQL) Create(key entity.Key) error {
-	statement := fmt.Sprintf(`
-INSERT INTO "%s" ("%s", "%s")
-VALUES ($1, $2);
-`,
-		table.AvailableKey.TableName,
-		table.AvailableKey.ColumnKey,
-		table.AvailableKey.ColumnCreatedAt,
-	)
+func (a AvailableKeySQL) Create(key entity.Key) (err error) {
+	return withTransactionSQL(a.db, func(tx *sql.Tx) error {
+		transactional := NewAvailableKeyTransactional(tx)
```

Comment:
I agree with you. `AvailableKeySQL` represents here `AvailableKeyTransactional` with `auto-commit` SQL mode. I've done the next:
* `AvailableKeySQL` is deleted
* `ProducerPersist` creates a bunch of keys in one transaction. From the performance perspective, it can be a useful change, but now we either create the whole bunch of keys or do not create anything (Here we should discuss it is ok or not)
* Now we have some 2 unpleasant (even ugly) `if conditions` [here](https://github.com/byliuyang/kgs/pull/43/files#diff-d71648324702f387df08f17a29306256R66-R68) and [here](https://github.com/byliuyang/kgs/pull/43/files#diff-d71648324702f387df08f17a29306256R78-R80). I don't know how to make it better, so if you have an idea, please let me know.

It appears to me, these new changes don't break any of `SOLID` principles.

Reply:
@alldroll The two `if` did look unnatural ( and error prone ). I can help experiment on the removing those `if` you can enable `allow edit from maintainers` option on the PR.

Reply:
Thanks, `Allow edits from maintainers` is enabled